### ISSUE TITLE
Arc Pro B support needs to be improved

### ISSUE THREAD

**Issue by pshirshov** [jamesbrink/nix-comfyui#56]
### Bug Description

# Two improvements from a downstream Battlemage (Arc Pro B70) deployment

I've been running `comfyui-nix` on an Arc Pro B70 (Xe2 / Battlemage / xe-kmd) host and hit two issues that needed downstream patches. Both have clean fixes that would benefit anyone running on Intel XPU.

---

### What the patch currently does

```diff
+ # Fallback to CPU if CUDA is not available and we're not on MPS
+ # This fixes running on Linux systems without NVIDIA GPUs
+ if cpu_state == CPUState.GPU:
+     if not torch.cuda.is_available():
+         cpu_state = CPUState.CPU
```

### The problem

The comment says "we're not on MPS", but the code only checks `torch.cuda.is_available()`. On any host where Intel XPU (or MPS) is the *only* GPU, this unconditionally clobbers `cpu_state` to `CPUState.CPU`, even though ComfyUI's own `is_intel_xpu()` would (correctly) return `True` immediately afterwards.

Observed symptom on Arc Pro B70 (`gpuSupport = "xpu"`):

* `torch.xpu.is_available()` → `True`
* `torch.xpu.device_count()` → `1`
* ComfyUI loads `ZImageTEModel_` (7.6 GB), `Lumina2` (23 GB) etc. **on CPU**
* `top` shows python @ 600–700% CPU, ~32 GB resident, **0% GPU activity** on `intel_gpu_top`

The `is_intel_xpu()` / `is_mps()` paths in upstream `model_management.py` never get a chance to set `cpu_state = GPU` because this CPU override comes after them.

### Proposed fix

Make the fallback backend-aware:

```diff
-# Fallback to CPU if CUDA is not available and we're not on MPS
-# This fixes running on Linux systems without NVIDIA GPUs
+# Fallback to CPU only if NO supported GPU backend is reachable.
+# Previously only CUDA was checked, which incorrectly clobbered cpu_state
+# to CPU on hosts with Intel XPU or Apple MPS.
 if cpu_state == CPUState.GPU:
-    if not torch.cuda.is_available():
+    if not (torch.cuda.is_available()
+            or (hasattr(torch, 'xpu') and torch.xpu.is_available())
+            or (hasattr(torch.backends, 'mps') and torch.backends.mps.is_available())):
         cpu_state = CPUState.CPU
```

Happy to file a PR with this change against the patch file.

---

## 2. `comfy-ui-xpu` does work on Battlemage, but only via the SYCL OpenCL UR adapter

This is more context-than-bug — useful for anyone trying the XPU package on a B-series Arc card.

### What we found

On Arc Pro B70 / kernel ≥ 7.0 / xe-kmd, NEO's Level Zero driver init aborts during the very first DRM allocation in `CommandStreamReceiver::initializeResources`:

```
Created new BO with GEM_USERPTR, handle: BO-0
Abort was called at 15 line in file:
../../neo/shared/source/gmm_helper/resource_info.cpp
```

Upstream bug: [intel/compute-runtime#922](https://github.com/intel/compute-runtime/issues/922). Same crash on `compute-runtime` `26.05`, `26.14`, `26.18`; same on `gmmlib` `22.9.0` and `22.10.0`; same on Intel's own `.deb`-distributed binaries. Confirmed independent of NEO/gmmlib version.

`torch.xpu` is therefore unusable through the default Level-Zero backend on Battlemage today.

### Workaround that works

SYCL's OpenCL UR adapter on the *same* NEO + same kernel succeeds. ollama-sycl, llama.cpp-SYCL, OpenVINO/ONNX-Runtime all already use it productively on this hardware.

PyTorch hardcodes the Level-Zero filter in [`c10/xpu/XPUFunctions.cpp:113`](https://github.com/pytorch/pytorch/blob/main/c10/xpu/XPUFunctions.cpp):

```cpp
if (platform.get_backend() != sycl::backend::ext_oneapi_level_zero) {
  return false;  // rejects OpenCL-backed SYCL platforms
}
```

It can be bypassed at runtime with a 1-symbol `LD_PRELOAD` shim (built `-nostdlib` so it doesn't pull glibc into the closure):

```cpp
extern "C" int _ZNK4sycl3_V18platform11get_backendEv(void*) { return 2; }
```

Combined with the env triple:

```
ONEAPI_DEVICE_SELECTOR=opencl:gpu
OCL_ICD_VENDORS=/run/opengl-driver/etc/OpenCL/vendors
LD_PRELOAD=<path-to-shim.so>
```

oneDNN's own backend switch in `dnnl::impl::gpu::intel::sycl::device_id` correctly handles the OpenCL path downstream, so the rest of `torch.xpu` works. Only `c10::xpu`'s pre-filter needed flipping.

### Validated end-to-end on Arc Pro B70 (32 GB ECC)

* `torch.xpu.device_count() = 1`, `torch.xpu.get_device_name(0) = "Intel(R) Graphics [0xe223]"`
* fp32 matmul 2048² @ **3.7 TFLOPS**
* fp16 1024² @ **3.2 TFLOPS** (cooperative-matrix path via `cl_intel_subgroup_matrix_multiply_accumulate_tf32`)
* `nn.Linear` training loop with Adam — loss descends as expected
* `F.scaled_dot_product_attention` on SD-UNet shapes (2×8×4096×64 fp16)
* Transformer block forward + backward + optimizer step

### Could this live in comfyui-nix?

Two options if interesting:

1. **Documentation only** — README note for B-series Arc users explaining the env triple + shim. Cheap.
2. **Bundled into the xpu package** — ship the `.so` in the xpu derivation and have `services.comfyui` (when `gpuSupport = "xpu"`) automatically set the env triple. Slightly more invasive but means it Just Works.

I'm happy to PR either path. Option 2 would need a way to detect Battlemage vs Alchemist at runtime (the shim is harmless on Alchemist but unnecessary), or just gate it on a new flag like `services.comfyui.xpu.openclBackend = true`.

### Steps to Reproduce

Attempt to run on Arc Pro B70 => pytorch crashes.

### Expected Behavior

-

### Error Logs

```shell

```

### Platform

Linux (x86_64)
